pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -59,17 +61,25 @@ type Plugin struct {
 
 	mutex                   sync.Mutex
 	conn                    *grpc.ClientConn
+	supportedAPI            apiVersion
```

Comment:
I thought the version that plugin tells kubelet is an API version it supports https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md#grpc-service-overview

So I think we are using the `highestSupportedVersion` incorrectly in DRA.

We can change the semantics of it in plugin registration API if this is what we want. I feel like plugin version is more interesting to kubelet than the versions of API it supports anyways and we can always use the approach with the fallback you are doing here

Reply:
What would be a better way of supporting a `v1beta1` and an older `v1alpha3`? Does it mean that during registration, a DRA plugin should announce which version it supports?

When I implemented that initially, I wasn't sure and left a TODO:
https://github.com/kubernetes/kubernetes/blame/d76a8fae674c972b5d2e2cfc226c607c9f63a412/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/noderegistrar.go#L40

I asked at that time, but didn't really get an answer, so this hasn't been addressed. I'll look into it.

Reply:
Yes, the idea was that plugin will tell which versions it supports and kubelet will pick which version to communicate to it by. So no fallbacks are needed on kubelet side - we just pick version we will use with this plugin when it registers

Reply:
BTW, this gives me troubles when I am implementing this: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4680-add-resource-health-to-pod-status/README.md#dra-implementation-details

If plugin only suppose to implement a single interface and gives back the version of that interface it supports, I am having issues understanding how to query for another interface from it. I will need a fallback logic in secondary interface or just add health-related methods to the existing interface

Reply:
Yes, optional features cannot be captured by a single version number.

Reply:
> When I implemented that initially, I wasn't sure and left a TODO:
https://github.com/kubernetes/kubernetes/blame/d76a8fae674c972b5d2e2cfc226c607c9f63a412/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/noderegistrar.go#L40

I looked at bit at how the kubelet side of things (I hadn't implemented that, so it's a bit new to me) and found a similar TODO here:
https://github.com/kubernetes/kubernetes/blob/c4eea34dcf6242602565e99426fda5f9183ca947/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go#L170-L181

That's exactly the situation that we are facing now.

For CSI plugins, this got turned into checking the CSI version:
https://github.com/kubernetes/kubernetes/blob/c4eea34dcf6242602565e99426fda5f9183ca947/pkg/volume/csi/csi_plugin.go#L162-L186

That code then got copied for DRA, without really defining what the version is supposed to mean. I should have followed up on that, sorry!

Using `HighestSupportedVersion` doesn't make sense for DRA, because a higher version (say, 1.33.0 for the next Kubernetes release) does not mean that the plugin supports all gRPC versions of a plugin supporting 1.32.0. For example, we might drop the alpha gRPC API. Besides, plugins might decide to not provide that to begin with!

I think we should list gRPC service names in that version slice, for example ["v1alpha3.Node"](https://github.com/kubernetes/kubernetes/blob/c4eea34dcf6242602565e99426fda5f9183ca947/staging/src/k8s.io/kubelet/pkg/apis/dra/v1alpha4/api.pb.go#L654). That is what matters to kubelet: it will know in advance what methods it can call because the plugin is serving them. This works betterif we are strict about versioning: we must not add new methods to existing versions, otherwise that "can be called" guarantee breaks down.

It might also make sense to rename `service Node` to `service DRAPlugin` in v1beta1. "Node" is not a very descriptive name and it stutters (`Node.NodePrepareResources`). I'll have to try how intrusive that'll be.

Does this make sense so far?

cc @klueska @bart0sh

Reply:
Regarding the renaming -- yes, I actually meant to propose changing that in the last API version bump but it got dropped. The name `DRAPlugin` is OK with me.

Reply:
> It might also make sense to rename service Node to service DRAPlugin in v1beta1. " [...]  I'll have to try how intrusive that'll be.

Not intrusive at all, at least for clients using the Go helper. See "DRA kubelet: rename gRPC server from Node to DRAPlugin in v1beta1".

> I think we should list gRPC service names in that version slice

Changed in "DRA kubelet: list supported gRPC services during registration". This replaces try-and-fall-back version selection with static selection based on the advertised gRPC services. In general, it's a major cleanup of versioning because that was also done inconsistently in tests.

@SergeyKanzhelev: can you take another look? You should be able to use this new convention for the health status reporting.